boards/BMS/Src/Io/Io_LTC6813.c

```diff
@@ -0,0 +1,374 @@
+    uint32_t received_pec15 =
+        (uint32_t)(rx_cell_voltages[cell_voltage_index] << 8) |
+        (uint32_t)(rx_cell_voltages[cell_voltage_index + 1]);
+    uint32_t calculated_pec15 = Io_CalculatePec15(
+        &rx_cell_voltages[current_ic * NUM_OF_CELL_VOLTAGE_RX_BYTES], 6U);
+
+    if (received_pec15 != calculated_pec15)
+    {
+        // Increment the number of PEC15 errors.
+        ltc_6813.pec15_error_counter++;
+    }
+}
+
+void Io_LTC6813_Init(
+    SPI_HandleTypeDef *hspi,
+    GPIO_TypeDef *     nss_port,
+    uint16_t           nss_pin)
+{
+    assert(hspi != NULL);
+
+    ltc_6813.spi                 = Io_SharedSpi_Create(hspi, nss_port, nss_pin);
+    ltc_6813.pec15_error_counter = 0U;
+    memset(
+        ltc_6813.cell_voltages, 0U,
+        NUM_OF_LTC6813 * NUM_OF_CELLS_PER_LTC6813 *
+            sizeof(ltc_6813.cell_voltages[0][0]));
+}
+
+void Io_LTC6813_Configure(void)
+{
+    uint8_t tx_cmd[NUM_OF_CMD_BYTES];
+    tx_cmd[0] = (uint8_t)(WRCFGA >> 8);
+    tx_cmd[1] = (uint8_t)WRCFGA;
+    uint16_t tx_cmd_pec15 =
+        Io_CalculatePec15(tx_cmd, NUM_OF_PEC15_BYTES_PER_CMD);
+    tx_cmd[2] = (uint8_t)(tx_cmd_pec15 >> 8);
+    tx_cmd[3] = (uint8_t)tx_cmd_pec15;
+
+    const uint32_t DEFAULT_CONFIG_REG[4] = {
+        (REFON << 2) + (DTEN << 1) + ADCOPT, (CELL_UNDERVOLTAGE_THRESHOLD),
+        (uint8_t)((CELL_OVERVOLTAGE_THRESHOLD & 0xF) << 4) +
+            (CELL_UNDERVOLTAGE_THRESHOLD >> 8),
+        (CELL_OVERVOLTAGE_THRESHOLD >> 4)
+    };
+
+    // The payload data is 8 bytes wide. The first 6 bytes is used to configure
+    // Configuration Register A, while the remaining two bytes are the PEC15 for
+    // the payload data transmitted.
+    uint8_t tx_payload[8] = { 0 };
+    memcpy(tx_payload, DEFAULT_CONFIG_REG, 4U);
+    uint16_t tx_payload_pec15 = Io_CalculatePec15(tx_payload, 6U);
+    tx_payload[6]             = (uint8_t)(tx_payload_pec15 >> 8);
+    tx_payload[7]             = (uint8_t)tx_payload_pec15;
+
+    Io_SharedSpi_SetNssLow(ltc_6813.spi);
+
+    // Transmit the command to write to Configuration Register A.
+    Io_SharedSpi_TransmitWithoutNssToggle(
+        ltc_6813.spi, tx_cmd, NUM_OF_CMD_BYTES);
+
+    // Transmit the payload data to all devices connected to the daisy chain.
+    Io_SharedSpi_MultipleTransmitWithoutNssToggle(
+        ltc_6813.spi, tx_payload, 8U, NUM_OF_LTC6813);
+    Io_SharedSpi_SetNssHigh(ltc_6813.spi);
+}
+
+ExitCode Io_LTC6813_ReadAllCellRegisterGroups(void)
+{
+    uint16_t cell_register_group_cmd;
+    uint8_t  tx_cmd[NUM_OF_CMD_BYTES];
+    uint8_t  rx_cell_voltages[NUM_OF_CELL_VOLTAGE_RX_BYTES * NUM_OF_LTC6813] = {
+        0
+    };
+
+    // Reset the value of the PEC15 error counter at the start of each cycle.
+    ltc_6813.pec15_error_counter = 0U;
+
+    RETURN_IF_EXIT_NOT_OK(Io_LTC6813_EnterReadyState())
+    RETURN_IF_EXIT_NOT_OK(Io_LTC6813_StartADCConversion())
+    RETURN_IF_EXIT_NOT_OK(Io_LTC6813_PollAdcConversion())
```

Comment:
This is a neat little macro :)

Reply:
Shout out to @markomarkusi

Reply:
At this point we might have to call him Macro Markusi